File: src/main/java/org/apache/fineract/client/services/HooksApi.java
/**
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements. See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership. The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License. You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied. See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */
package org.apache.fineract.client.services;

import org.apache.fineract.client.CollectionFormats.*;

import rx.Observable;
import retrofit2.http.*;

import okhttp3.RequestBody;
import okhttp3.ResponseBody;
import okhttp3.MultipartBody;

import org.apache.fineract.client.models.DeleteHookResponse;
import org.apache.fineract.client.models.GetHookResponse;
import org.apache.fineract.client.models.GetHookTemplateResponse;
import org.apache.fineract.client.models.PostHookRequest;
import org.apache.fineract.client.models.PostHookResponse;
import org.apache.fineract.client.models.PutHookRequest;
import org.apache.fineract.client.models.PutHookResponse;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

public interface HooksApi {
  /**
   * Create a Hook
   * The following parameters can be passed for the creation of a hook :-  name - string - Required. The name of the template that is being called. (See /hooks/template for the list of valid hook names.)  isActive - boolean - Determines whether the hook is actually triggered.  events - array - Determines what events the hook is triggered for.  config - hash - Required. Key/value pairs to provide settings for this hook. These settings vary between the templates.  templateId - Optional. The UGD template ID associated with the same entity (client or loan).
   * @param postHookRequest  (required)
   * @return Observable&lt;PostHookResponse&gt;
   */
  @Headers({
    "Content-Type:application/json"
  })
  @POST("hooks")
  Observable<PostHookResponse> createHook(
    @retrofit2.http.Body PostHookRequest postHookRequest
  );

  /**
   * Delete a Hook
   * Deletes a hook.
   * @param hookId hookId (required)
   * @return Observable&lt;DeleteHookResponse&gt;
   */
  @DELETE("hooks/{hookId}")
  Observable<DeleteHookResponse> deleteHook(
    @retrofit2.http.Path("hookId") Long hookId
  );

  /**
   * Retrieve a Hook
   * Returns the details of a Hook.  Example Requests:  hooks/1
   * @param hookId hookId (required)
   * @return Observable&lt;GetHookResponse&gt;
   */
  @GET("hooks/{hookId}")
  Observable<GetHookResponse> retrieveHook(
    @retrofit2.http.Path("hookId") Long hookId
  );

  /**
   * Retrieve Hooks
   * Returns the list of hooks.  Example Requests:  hooks
   * @return Observable&lt;List&lt;GetHookResponse&gt;&gt;
   */
  @GET("hooks")
  Observable<List<GetHookResponse>> retrieveHooks();
    

  /**
   * Retrieve Hooks Template
   * This is a convenience resource. It can be useful when building maintenance user interface screens for client applications. The template data returned consists of any or all of:  Field Defaults Allowed description Lists Example Request:  hooks/template
   * @return Observable&lt;GetHookTemplateResponse&gt;
   */
  @GET("hooks/template")
  Observable<GetHookTemplateResponse> template3();
    

  /**
   * Update a Hook
   * Updates the details of a hook.
   * @param hookId hookId (required)
   * @param putHookRequest  (required)
   * @return Observable&lt;PutHookResponse&gt;
   */
  @Headers({
    "Content-Type:application/json"
  })
  @PUT("hooks/{hookId}")
  Observable<PutHookResponse> updateHook(
    @retrofit2.http.Path("hookId") Long hookId, @retrofit2.http.Body PutHookRequest putHookRequest
  );

}

File: src/main/kotlin/org/openapitools/client/apis/HooksApi.kt
package org.openapitools.client.apis

import org.openapitools.client.models.DeleteHookResponse
import org.openapitools.client.models.GetHookResponse
import org.openapitools.client.models.GetHookTemplateResponse
import org.openapitools.client.models.PostHookRequest
import org.openapitools.client.models.PostHookResponse
import org.openapitools.client.models.PutHookRequest
import org.openapitools.client.models.PutHookResponse
import retrofit2.http.Body
import retrofit2.http.DELETE
import retrofit2.http.GET
import retrofit2.http.POST
import retrofit2.http.PUT
import retrofit2.http.Path

interface HooksApi {
    /**
     * Create a Hook
     * The following parameters can be passed for the creation of a hook :-  name - string - Required. The name of the template that is being called. (See /hooks/template for the list of valid hook names.)  isActive - boolean - Determines whether the hook is actually triggered.  events - array - Determines what events the hook is triggered for.  config - hash - Required. Key/value pairs to provide settings for this hook. These settings vary between the templates.  templateId - Optional. The UGD template ID associated with the same entity (client or loan).
     * Responses:
     *  - 200: OK
     *
     * @param postHookRequest
     * @return [PostHookResponse]
     */
    @POST("v1/hooks")
    suspend fun createHook(@Body postHookRequest: PostHookRequest): PostHookResponse

    /**
     * Delete a Hook
     * Deletes a hook.
     * Responses:
     *  - 200: OK
     *
     * @param hookId hookId
     * @return [DeleteHookResponse]
     */
    @DELETE("v1/hooks/{hookId}")
    suspend fun deleteHook(@Path("hookId") hookId: Long): DeleteHookResponse

    /**
     * Retrieve a Hook
     * Returns the details of a Hook.  Example Requests:  hooks/1
     * Responses:
     *  - 200: OK
     *
     * @param hookId hookId
     * @return [GetHookResponse]
     */
    @GET("v1/hooks/{hookId}")
    suspend fun retrieveHook(@Path("hookId") hookId: Long): GetHookResponse

    /**
     * Retrieve Hooks
     * Returns the list of hooks.  Example Requests:  hooks
     * Responses:
     *  - 200: OK
     *
     * @return [kotlin.collections.List<GetHookResponse]
     */
    @GET("v1/hooks")
    suspend fun retrieveHooks(): List<GetHookResponse>

    /**
     * Retrieve Hooks Template
     * This is a convenience resource. It can be useful when building maintenance user interface screens for client applications. The template data returned consists of any or all of:  Field Defaults Allowed description Lists Example Request:  hooks/template
     * Responses:
     *  - 200: OK
     *
     * @return [GetHookTemplateResponse]
     */
    @GET("v1/hooks/template")
    suspend fun template3(): GetHookTemplateResponse

    /**
     * Update a Hook
     * Updates the details of a hook.
     * Responses:
     *  - 200: OK
     *
     * @param hookId hookId
     * @param putHookRequest
     * @return [PutHookResponse]
     */
    @PUT("v1/hooks/{hookId}")
    suspend fun updateHook(
        @Path("hookId") hookId: Long,
        @Body putHookRequest: PutHookRequest
    ): PutHookResponse

}
